### PR TITLE
Issue #143: Optimize atomium_get_settings().

### PR DESCRIPTION
**Pull request by drupol**

### CODE REVIEW

File: atomium/includes/common.inc [hunk 1]
```diff
@@ -675,33 +675,141 @@ function _atomium_array_flatten(array $array, array &$storage, $parentKey = '')
...
+ *   Callbacks to apply to setting values.
+ *
+ * @return mixed
+ *   The settings.
+ *
+ * @SuppressWarnings(PHPMD.CyclomaticComplexity)
+ * @SuppressWarnings(PHPMD.NPathComplexity)
+ */
+function _atomium_get_settings_flat_filtered($theme_key, $base_themes = TRUE, array $value_callbacks = array('trim')) {
+  static $settings_flat_filtered_by_cid = array();
```
Comment (imanoleguskiza) [inline atomium/includes/common.inc]: Why not use drupal_static?

**Review by imanoleguskiza — DISMISSED** [review]
Could we have information on how much performance gain this PR would bring? It seems to me we are trading a bit of performance for much more simpler code.

**Comment by donquixote** [comment]
@imanoleguskiza The performance impact is significant.
This was repeated a lot on average page load. Now it runs only once.

From what I remember in my profiling, it may reduce page load time by 5% or 10%.. but depends how you measure.

@drupol Can you check with your apache benchmarks?

File: atomium/includes/common.inc (hunk 1 above)
Comment (donquixote) [reply atomium/includes/common.inc]: The benefit of `drupal_static()` would be that other code (e.g. unit tests) would be able to reset this cache. Is someone planning to do this?

If we use `drupal_static()`, we should also use the drupal_static_fast() pattern I guess..

**Review by drupol — COMMENTED** [review]
I see that there's a lot of code change... and it's a bit harder to understand the logic (_especially with the `(t)` and `(b)` strings_).

I do believe you when you say that performances are better but... wouldn't it be easier to use it how it was and statically cache the result to gain performance ?

**Comment by donquixote** [comment]
If you want fewer code changes, look at the first of the two commits.
https://github.com/ec-europa/atomium/pull/201/commits/86ce1072f09a8c5f85005bee56f4497b726a45dd

That one only prevents `_atomium_array_flatten()` from being called repeatedly.

**Comment by donquixote** [comment]
> If you want fewer code changes, look at the first of the two commits.

Both commits make a performance improvement.

@Pol can you run your apache benchmarks on
- original
- first commit
- second commit

**Comment by drupol** [comment]
Here's the results with Apache benchmarks.

The command I used is:

`ab -n 1000 -kc 10 http://127.0.0.1:8080/build/node/1`

This is the branch 7.x-2.x without modifications: [atomium-bartik-7.x-2.x.log](https://github.com/ec-europa/atomium/files/2653069/atomium-bartik-7.x-2.x.log)

This is the @donquixote 's branch 143 as it is now: [atomium-bartik-donquixote-143-7.x-2.x.log](https://github.com/ec-europa/atomium/files/2653072/atomium-bartik-donquixote-143-7.x-2.x.log)

This is the branch 7.x-2.x optimized with `drupal_static_fast` (_this is a test that is not committed yet_)
[atomium-bartik-optimized-7.x-2.x.log](https://github.com/ec-europa/atomium/files/2653071/atomium-bartik-optimized-7.x-2.x.log)

**Comment by drupol** [comment]
I made further tests, including a `sleep(2)` in `atomium_get_settings()` ... and it does the job.
The first request is much longer than the next ones.

However, I get better results using this:
```php
function atomium_get_settings($setting_keys, $base_themes = TRUE) {
  static $drupal_static_fast;

  if (!isset($drupal_static_fast)) {
    $drupal_static_fast['atomium_get_settings'] = &drupal_static(__FUNCTION__);
  }

  $theme_key = $GLOBALS['theme_key'];
  $cid = $theme_key;
  if (TRUE === $base_themes) {
    $cid .= '+';
  }

  $atomium_get_settings = &$drupal_static_fast['atomium_get_settings'];

  if (isset($atomium_get_settings[$cid])) {
    return $atomium_get_settings[$cid];
  }

  $settings = array();

  if ($theme_settings = atomium_get_theme_info($theme_key, 'settings', $base_themes)) {
    _atomium_array_flatten(
      $theme_settings,
      $settings
    );
  }

  $setting_value = array();

  if (isset($settings[$setting_keys])) {
    if (\is_string($settings[$setting_keys])) {
      $setting_value = \explode(',', $settings[$setting_keys]);
    }

    if (\is_array($settings[$setting_keys])) {
      $setting_value = $settings[$setting_keys];
    }

    $setting_value = \array_map('trim', $setting_value);
  }

  $atomium_get_settings[$cid] = \array_values(\array_filter($setting_value));

  return $atomium_get_settings[$cid];
}
```

**Review by drupol — COMMENTED** [review]
I'd use your first commit and update it to use drupal_static_fast, just like in the comment I did. This is the best and cleanest solution I think.

**Comment by donquixote** [comment]
@drupol
I would like to do some profiling myself.
The results imply that my branch actually is slower, so I want to understand what is happening there.
We also need to make sure that the page we test is actually affected in a relevant way by the changes, and this is not just coincidence.

Would it be ok if I do that next week?

**Comment by drupol** [comment]
> Would it be ok if I do that next week?

Sure mate, please do, it will be nice to have your input as well.

**Review by richardcanoe — APPROVED** [review]

**Comment by donquixote** [comment]
Hi,
I tried to do some profiling for this. I tried this with a local EMA instance I had for this same purpose (until I attempted to update it, which was not a good idea, and with a local test site.
I am a bit disappointed by my own results.
The improvement is less than what I remember it to be in older experiments.

I think the performance impact depends on two things:
- How many times atomium_get_settings() is called on the given page.
- How many settings are defined in the sub-theme, and how deeply they are nested.

## How often does atomium_get_settings() get called?

atomium_field_attach_view_alter()
atomium_preprocess_block()
atomium_preprocess_node()
atomium_preprocess_region()

I remember xdebug profiling results where this was called 100x or more.

But in the tests I am running now, I don't get these high numbers.

I imagine we can get high numbers on pages like this:
- Pages with many nodes displayed at once (e.g. a huge view, perhaps where some nodes have ).
- A view with a lot of views fields, where hook_field_attach_view_alter() would be called for each field on each row.

I do not have such examples easily available atm.

## How many settings can a sub-theme define?
It seems most sub-themes only define a limited number of settings for atomium.

## What are the profiling results?
I had a local test site where the first commit in my own PR would 

## Profiling methods
I have 4 methods currently available to profile this stuff:
- Measure page load time with the browser Network tab.
- Measure page load time with apache bench. This is good to test concurrent requests.
- Measure page load time with cfrprofiler "Curl request". This is good to reduce statistical variation.
- Measure the time for specific operations (atomium_get_settings()) directly with cfrprofiler. This is a bit tricky because we cannot easily reset the static cache between runs. Instead I am running another piece of code that does the same but outside of Atomium.

Page requests generally have a high statistical variation, especially on a developer machine where other unrelated processes run at the same time. Repetition is the only way to mitigate this.

## Profiling results
I can, with enough repetition, confirm that the first of the commits brings an improvement. The improvement of the second commit is too hard to measure. The last commit by Pol is very similar.

cfrprofiler page requests on a local test site, page load time, best of 20:
- Original 7.x-2.x 2d53af88: Most results between 104ms - 110ms, but I saw one 100ms result.
- With the first commit 86ce1072: 103ms - 108ms.
- With the second commit 8bd92994: Similar results, too much variation to pin it down.
- Last commit 8bd92994 by Pol: Similar results, too much variation to pin it down.

Artificially increasing the number of calls to atomium_get_settings() makes the improvement from the first commit more visible.

cfrprofiler simulated algorithm: This shows that the optimized version from the 1st request 10x faster than the non-optimized version, on repeated calls.
The second commit brings a tiny extra improvement if measured this way, but it is really hard to confirm this with statistical significance.

## Conclusion

I agree now that the second commit is not worth the added complexity.

I also agree to use drupal_static / drupal_static_fast.

The commit by Pol also removes the parameter $value_callbacks, and behaves as if it has the default value. I tried not to do this in 7.x-2.x for BC reasons, but it is Pol's choice as a maintainer to do this change for the sake of simplicity.

I just think such a change should be in a dedicated commit.
I also think the git history will become simpler if we omit the second commit which introduces the extra changes.

I am going to push something with a simpler history.

**Comment by donquixote** [comment]
@drupol I just noticed a severe bug in the new commit.
You return `$atomium_get_settings[$cid]`, but the `$cid` does not include the `$setting_keys`.

The result, on atomium_bartik:

```php
dpm(atomium_get_settings('alter.css_alter.exclude'));
dpm(atomium_get_settings('toggle_favicon'));
```
This gives me 2x the full array of alter.css_alter.exclude.

```php
dpm(atomium_get_settings('alter.css_alter.exclude'));
dpm(atomium_get_settings('toggle_favicon'));
```
This gives me 2x the value 1 of toggle_favicon.

**Comment by donquixote** [comment]
@drupol I pushed a tag where I split the history and omit my own 2nd commit.
https://github.com/ec-europa/atomium/compare/7.x-2.x...143-7x2x-atomium_get_settings-optimize-SPLIT
(I don't want to force-push to this PR, but also don't want to create yet another PR).
The end commit there has the same code as your end commit.

Imo we only need the first commit in this sequence, 79d7fac0c247a87e4940b6885e0854da9ed163cd
So we would end up with this:
https://github.com/ec-europa/atomium/compare/7.x-2.x...79d7fac0c247a87e4940b6885e0854da9ed163cd

No BC break, no new bugs (unless you find one), and we can still do other things as follow-up if we want.

**Comment by drupol** [comment]
>  You return `$atomium_get_settings[$cid]`, but the `$cid` does not include the `$setting_keys`.

You're right. The thing is that, do you think we will have more than one call with the same signature?

First request: atomium_get_settings('alter.css_alter.exclude'); // Not cached
Second request: atomium_get_settings('alter.css_alter.exclude'); // Cached

If you think that this would be useful, then, I propose to generate the CID using:

```php
  $cid_parameters = \func_get_args();
  $cid_parameters['theme_key'] = $GLOBALS['theme_key'];
  $cid = \sha1(\json_encode($cid_parameters));
```

What do you think ?

**Comment by drupol** [comment]
> @drupol I pushed a tag where I split the history and omit my own 2nd commit.
> [7.x-2.x...143-7x2x-atomium_get_settings-optimize-SPLIT](https://github.com/ec-europa/atomium/compare/7.x-2.x...143-7x2x-atomium_get_settings-optimize-SPLIT)
> (I don't want to force-push to this PR, but also don't want to create yet another PR).
> The end commit there has the same code as your end commit.

Thanks for redoing it, it's perfectly fine and I do not mind if you force push on my branch, do not hesitate to push force on this branch.

**Comment by donquixote** [comment]
You have multiple calls with different setting keys. As many as there are
settings in the subtheme. Each time you would run the array flattening.

Am Do., 13. Dez. 2018, 09:50 hat Pol Dellaiera <notifications@github.com>
geschrieben:

**Comment by drupol** [comment]
> You have multiple calls with different setting keys. As many as there are settings in the subtheme. Each time you would run the array flattening.

How about then to use `drupal_static_fast` on `atomium_get_theme_info()` and `_atomium_array_flatten()` as well ?

**Comment by donquixote** [comment]
We are then getting back to my original PR which you found too complicated
😊

Am Do., 13. Dez. 2018, 10:19 hat Pol Dellaiera <notifications@github.com>
geschrieben:

**Comment by drupol** [comment]
It's a bit different, you were using two new functions as far as I remember.

What I propose here is to use the drupal_static_fast on those functions so they can be used by other process as well.

**Comment by donquixote** [comment]
> How about then to use drupal_static_fast on atomium_get_theme_info() and _atomium_array_flatten() as well ?

> What I propose here is to use the drupal_static_fast on those functions so they can be used by other process as well.

For atomium_get_theme_info() this can work, because this has only few parameter variations. However, this is only called in very few places. Once `atomium_get_settings()` gets the static cache, it will run 1x per request.

For _atomium_array_flatten() I have to say no, because the entire array of non-flattened theme settings would become the cache id.

Imo we should go with the simple and low-risk version https://github.com/ec-europa/atomium/compare/7.x-2.x...79d7fac0c247a87e4940b6885e0854da9ed163cd, and leave whatever else we want to do as a follow-up.

**Comment by donquixote** [comment]
> Imo we should go with the simple and low-risk version https://github.com/ec-europa/atomium/compare/7.x-2.x...79d7fac0c247a87e4940b6885e0854da9ed163cd, and leave whatever else we want to do as a follow-up.

We are talking about 7.x-2.x branch after all, so low-risk + easy win is the way to go.

**Comment by donquixote** [comment]
@drupol 
As discussed, I force-pushed the simplified branch on this PR.
It contains the first commit of the original PR, and one small commit to use the drupal_static_fast pattern.

You can create a local archive tag or whatever you like to not lose your own version of the branch.

**Comment by drupol** [comment]
All good for me now, I approve this.

**Comment by drupol** [comment]
Thanks, cherry picked in 7.x-3.x.

**Pull request merged**